File: vantare-v2/frontend/src/lib/plan.test.ts
import { describe, expect, it } from "vitest";
import {
  buildSummary,
  classifyPlan,
  classifyStatus,
  sortedEntitlements,
  PLAN_LABELS,
  PLAN_STATUS_LABELS,
} from "./plan";
import type { Entitlement, LicenseState } from "./license-types";

describe("classifyPlan", () => {
  const cases: Array<{ name: string; in: Entitlement[]; want: string }> = [
    { name: "empty is free", in: [], want: "free" },
    { name: "overlays alone", in: ["overlays"], want: "paid_overlays" },
    { name: "engineer alone", in: ["engineer"], want: "paid_engineer" },
    { name: "bundle is suite", in: ["bundle"], want: "suite" },
    { name: "beta_access is suite", in: ["beta_access"], want: "suite" },
    { name: "founder is suite", in: ["founder"], want: "suite" },
    { name: "pro_founder is suite", in: ["pro_founder"], want: "suite" },
    { name: "visionary_backer is suite", in: ["visionary_backer"], want: "suite" },
    { name: "supporter is paid_overlays", in: ["supporter"], want: "paid_overlays" },
    { name: "bundle plus ac_lua_pack is suite", in: ["bundle", "ac_lua_pack"], want: "suite" },
    { name: "ac_lua_pack alone is free", in: ["ac_lua_pack"], want: "free" },
    {
      name: "overlays plus engineer is suite",
      in: ["overlays", "engineer"],
      want: "suite",
    },
    { name: "unknown token", in: ["mystery_key" as Entitlement], want: "unknown" },
    { name: "whitespace ignored", in: ["" as Entitlement], want: "free" },
  ];

  for (const tc of cases) {
    it(tc.name, () => {
      expect(classifyPlan(tc.in)).toBe(tc.want);
    });
  }
});

describe("classifyStatus", () => {
  const states: Array<[LicenseState | null, string]> = [
    ["active", "active"],
    ["grace", "grace"],
    ["expired", "blocked"],
    ["device-limit", "blocked"],
    ["authenticated-no-entitlement", "free"],
    ["anonymous", "anonymous"],
    ["unconfigured", "unconfigured"],
    [null, "free"],
  ];

  for (const [state, expected] of states) {
    it(`${String(state)} -> ${expected}`, () => {
      expect(classifyStatus(state)).toBe(expected);
    });
  }
});

describe("buildSummary", () => {
  it("active suite", () => {
    expect(buildSummary("active", ["bundle"])).toEqual({
      label: "suite",
      status: "active",
    });
  });
  it("grace overlays", () => {
    expect(buildSummary("grace", ["overlays"])).toEqual({
      label: "paid_overlays",
      status: "grace",
    });
  });
  it("blocked without entitlements", () => {
    expect(buildSummary("expired", [])).toEqual({
      label: "free",
      status: "blocked",
    });
  });
  it("authenticated-no-entitlement is free", () => {
    expect(buildSummary("authenticated-no-entitlement", [])).toEqual({
      label: "free",
      status: "free",
    });
  });
  it("anonymous forces anonymous status even with suite entitlements", () => {
    expect(buildSummary("anonymous", ["bundle"])).toEqual({
      label: "suite",
      status: "anonymous",
    });
  });
  it("active unknown becomes free", () => {
    expect(buildSummary("active", ["mystery" as Entitlement])).toEqual({
      label: "free",
      status: "active",
    });
  });
  it("unconfigured is not blocked", () => {
    expect(buildSummary("unconfigured", [])).toEqual({
      label: "free",
      status: "unconfigured",
    });
  });
});

describe("sortedEntitlements", () => {
  it("returns a sorted copy without mutating the input", () => {
    const src: Entitlement[] = ["bundle", "overlays", "engineer"];
    const got = sortedEntitlements(src);
    expect(got).toEqual(["bundle", "engineer", "overlays"]);
    expect(src).toEqual(["bundle", "overlays", "engineer"]);
  });
});

describe("label maps", () => {
  it("covers every plan label", () => {
    const labels: string[] = ["free", "paid_overlays", "paid_engineer", "suite", "unknown"];
    for (const label of labels) {
      expect(PLAN_LABELS[label as keyof typeof PLAN_LABELS]).toBeTruthy();
    }
  });
  it("covers every plan status", () => {
    const statuses: string[] = ["active", "grace", "blocked", "free", "anonymous", "unconfigured"];
    for (const status of statuses) {
      expect(PLAN_STATUS_LABELS[status as keyof typeof PLAN_STATUS_LABELS]).toBeTruthy();
    }
  });
});

File: vantare-v2/internal/license/plan_test.go
package license

import "testing"

// TestClassifyPlan is table-driven and locked to docs/stripe-integration-plan.md.
func TestClassifyPlan(t *testing.T) {
	cases := []struct {
		name string
		in   []Entitlement
		want PlanLabel
	}{
		{"empty is free", nil, PlanFree},
		{"empty slice is free", []Entitlement{}, PlanFree},
		{"overlays alone", []Entitlement{EntitlementOverlays}, PlanPaidOverlays},
		{"engineer alone", []Entitlement{EntitlementEngineer}, PlanPaidEngineer},
		{"bundle is suite", []Entitlement{EntitlementBundle}, PlanSuite},
		{"beta_access is suite", []Entitlement{EntitlementBetaAccess}, PlanSuite},
		{"founder is suite", []Entitlement{EntitlementFounder}, PlanSuite},
		{"pro_founder is suite", []Entitlement{EntitlementProFounder}, PlanSuite},
		{"visionary_backer is suite", []Entitlement{EntitlementVisionaryBacker}, PlanSuite},
		{"supporter is paid_overlays", []Entitlement{EntitlementSupporter}, PlanPaidOverlays},
		{"bundle plus ac_lua_pack is suite", []Entitlement{EntitlementBundle, EntitlementACLuaPack}, PlanSuite},
		{"ac_lua_pack alone is free add-on", []Entitlement{EntitlementACLuaPack}, PlanFree},
		{"overlays plus engineer is suite (treated as bundle)", []Entitlement{EntitlementOverlays, EntitlementEngineer}, PlanSuite},
		{"unknown token", []Entitlement{"mystery_key"}, PlanUnknown},
		{"whitespace ignored", []Entitlement{""}, PlanFree},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := ClassifyPlan(tc.in)
			if got != tc.want {
				t.Fatalf("ClassifyPlan(%v) = %q, want %q", tc.in, got, tc.want)
			}
		})
	}
}

// TestClassifyStatus locks the State -> PlanStatus mapping.
func TestClassifyStatus(t *testing.T) {
	cases := []struct {
		in   State
		want PlanStatus
	}{
		{StateActive, PlanStatusActive},
		{StateGrace, PlanStatusGrace},
		{StateExpired, PlanStatusBlocked},
		{StateDeviceLimit, PlanStatusBlocked},
		{StateAuthenticatedNoEntitlement, PlanStatusFree},
		{StateAnonymous, PlanStatusAnonymous},
		{StateUnconfigured, PlanStatusUnconfigured},
		{"other", PlanStatusFree},
	}

	for _, tc := range cases {
		t.Run(string(tc.in), func(t *testing.T) {
			got := ClassifyStatus(tc.in)
			if got != tc.want {
				t.Fatalf("ClassifyStatus(%q) = %q, want %q", tc.in, got, tc.want)
			}
		})
	}
}

func TestBuildSummary(t *testing.T) {
	cases := []struct {
		name         string
		state        State
		entitlements []Entitlement
		wantLabel    PlanLabel
		wantStatus   PlanStatus
	}{
		{
			name:         "active suite",
			state:        StateActive,
			entitlements: []Entitlement{EntitlementBundle},
			wantLabel:    PlanSuite,
			wantStatus:   PlanStatusActive,
		},
		{
			name:         "grace overlays",
			state:        StateGrace,
			entitlements: []Entitlement{EntitlementOverlays},
			wantLabel:    PlanPaidOverlays,
			wantStatus:   PlanStatusGrace,
		},
		{
			name:         "blocked without entitlements",
			state:        StateExpired,
			entitlements: nil,
			wantLabel:    PlanFree,
			wantStatus:   PlanStatusBlocked,
		},
		{
			name:         "authenticated-no-entitlement is free",
			state:        StateAuthenticatedNoEntitlement,
			entitlements: nil,
			wantLabel:    PlanFree,
			wantStatus:   PlanStatusFree,
		},
		{
			name:         "blocked suite loses suite label",
			state:        StateDeviceLimit,
			entitlements: []Entitlement{EntitlementBundle},
			wantLabel:    PlanSuite,
			wantStatus:   PlanStatusBlocked,
		},
		{
			name:         "anonymous forces anonymous status",
			state:        StateAnonymous,
			entitlements: []Entitlement{EntitlementBundle},
			wantLabel:    PlanSuite,
			wantStatus:   PlanStatusAnonymous,
		},
		{
			name:         "active unknown becomes free",
			state:        StateActive,
			entitlements: []Entitlement{"mystery"},
			wantLabel:    PlanFree,
			wantStatus:   PlanStatusActive,
		},
		{
			name:         "unconfigured is not blocked",
			state:        StateUnconfigured,
			entitlements: nil,
			wantLabel:    PlanFree,
			wantStatus:   PlanStatusUnconfigured,
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := BuildSummary(tc.state, tc.entitlements)
			if got.Label != tc.wantLabel {
				t.Fatalf("label = %q, want %q", got.Label, tc.wantLabel)
			}
			if got.Status != tc.wantStatus {
				t.Fatalf("status = %q, want %q", got.Status, tc.wantStatus)
			}
		})
	}
}

func TestSortedEntitlements(t *testing.T) {
	got := SortedEntitlements([]Entitlement{"bundle", "overlays", "engineer"})
	want := []Entitlement{"bundle", "engineer", "overlays"}
	if len(got) != len(want) {
		t.Fatalf("len = %d, want %d", len(got), len(want))
	}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("idx %d: got %q, want %q", i, got[i], want[i])
		}
	}

	// Input is not mutated.
	src := []Entitlement{"b", "a"}
	_ = SortedEntitlements(src)
	if src[0] != "b" || src[1] != "a" {
		t.Fatalf("input mutated: %v", src)
	}
}
